H-2194, H-1777: add `/admin/users` page where hash instance admins can set the feature flags of a user

benwerner01 opened the pull request:
## 🌟 What is the purpose of this PR?

<!-- Explain, at a high level, what this does and why. -->
<!-- Use the 'What does this change?' section to list more specific implementation details. -->

This PR adds `/admin/users` and `/admin/users/:uuid` pages, so that the enabled feature flags of users can be modified by HASH instance admins. The `/admin/user` page can be navigated to via the sidebar and the account dropdown (H-1777)

## 🔗 Related links

<!-- Add links to any context it is worth capturing (e.g. Issues, Discussions, Discord) -->
<!-- Mark any links which are not publicly accessible as _(internal)_ -->
<!-- Don't rely on links to explain the PR, especially internal ones: use the sections above -->

- H-2194
- H-1777

## 🔍 What does this change?

<!-- Use a bullet list to explain your changes in more detail, if it would be helpful. -->
<!-- If applicable, link to the specific commit.-->

- adds a migration which gives the instance admin account group edit permissions on all existing user entities
- when creating a new user, gives the instance admin account group edit permissions on the new user entity

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?

<!-- Confirm you have taken the necessary action to record a changeset or publish a change, as appropriate -->
<!-- Tick AT LEAST ONE box and delete the rest. Do not delete this section! see libs/README.md for info on publishing -->

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

<!-- If this adds or moves an existing package, modifies `scripts` in a `package.json`, it likely needs a turbo graph change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->

## 🐾 Next steps

<!-- Are there are planned/suggested follow ups which are related but won't be done in this PR? -->

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

Manual testing

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->

1. Login as an instance admin
2. Navigate to `/admin/users` via the sidebar or the account dropdown
3. Modify the enabled feature flags for an existing user
4. Ensure they persisted correctly (by refreshing the page for example)

## 📹 Demo

<!-- Add a screenshot or video showcasing your work -->

https://github.com/hashintel/hash/assets/42802102/def23b9a-46b1-49c9-9b12-59f3eb496421

CiaranMn reviewed (commented):
One comment on permissions – can we also direct non-admin users away from the `/admin` route somehow?

benwerner01 commented on apps/hash-api/src/graph/knowledge/system-types/user.ts:
```diff
@@ -435,6 +439,27 @@ export const createUser: ImpureGraphFunction<
+    await getHashInstanceAdminAccountGroupId(ctx, authentication);
+
+  /** Grant permissions to hash instance admins to edit the user entity */
+  await modifyEntityAuthorizationRelationships(ctx, authentication, [
```
Good point, have addressed in https://github.com/hashintel/hash/pull/4203/commits/c659cc62d2c4de80ac2c1bb8a67661f174201cf6.

> Something that is already attempting to do this is further above in the code, but I suspect it has broken since the HASH Instance entity is no longer owned by the HASH instance account group (it's owned by the HASH org) – should be fixable by getting the accountGroupId how you are here.

I fixed this in the same commit. I assume this means there might be some user entities in production where the HASH instance admin group doesn't have the `administrator` permission, if so is this something you'd like me to address in the migration? Or ticketize?

CiaranMn replied:
We don't need both `editor` and `administrator`, because `administrator` implies `editor` – I'd just give `administrator`

benwerner01 replied:
Ah nice, then there isn't an issue. Have addressed in https://github.com/hashintel/hash/pull/4203/commits/72509f53af92c8ace1b1cb063ca0f0541b024e0a and tested it out locally 👍

benwerner01 commented:
@nonparibus 

> Can we show "email address" in the column, as well? Many users without shortnames aren't visible right now, so it's not easy to upgrade.

Have added the email address in https://github.com/hashintel/hash/pull/4203/commits/bdf1c0115fe72c233de41a63402ee19338f05b6d. When we start hiding user email addresses (H-2231), we'll need to make sure that we account for HASH instance admins not losing visibility in the process (cc @CiaranMn)


> We should redirect instance admins from /admin to /admin/users if there is nothing available there.

I've added a redirect from `/admin` to `/admin/user` in https://github.com/hashintel/hash/pull/4203/commits/cfd163119f49bb370cd3472f9e6e9f6ba8efd6f3, similar to what we're doing on the `/settings` page right now. The navigation links now point to `/admin` as well, so we don't forget to change them once we do have something hosted on the `/admin` page.

> We should redirect non-instance admins from /admin to the homepage, who do not have permission to view it.

I've addressed this in https://github.com/hashintel/hash/pull/4203/commits/2e5a7aa8a27d89dc76836cbbd40e4d7c1c6c8480.

CiaranMn commented on apps/hash-frontend/src/pages/admin/admin-page-layout.tsx:
```diff
@@ -39,6 +45,21 @@ const AdminLayout = ({ children }: PropsWithChildren) => {
+    /**
+     * Redirect non instance admins away from `/admin` pages
+     */
+    if (typeof isInstanceAdmin === "boolean" && !isInstanceAdmin) {
```
Nit, can merge it as it is
```suggestion
    if (isInstanceAdmin === false) {
```

CiaranMn reviewed (approved):
Good to merge, one minor simplification opportunity I noticed that needn't block merging
